javalin/src/main/java/io/javalin/http/staticfiles/PrecompressingResourceHandler.kt

```diff
@@ -0,0 +1,112 @@
+import javax.servlet.http.HttpServletRequest
+import javax.servlet.http.HttpServletResponse
+
+
+class PrecompressingResourceHandler(val compStrat: CompressionStrategy) {
+
+    val compressedFiles = HashMap<String, ByteArray>()
+
+    companion object {
+        val excludedMimeTypes = setOf(
+                "image/",
+                "audio/",
+                "video/",
+                "application/compress",
+                "application/zip",
+                "application/gzip",
+                "application/bzip2",
+                "application/brotli",
+                "application/x-xz",
+                "application/x-rar-compressed")
+    }
+
+    fun handle(resource: Resource, req: HttpServletRequest, res: HttpServletResponse): Boolean {
+        if (resource.exists() && !resource.isDirectory) {
+            val target = req.getAttribute("jetty-target") as String
+            var acceptCompressType = CompressType.getByAcceptEncoding(req.getHeader(Header.ACCEPT_ENCODING) ?: "")
+            val contentType = MimeTypes.getDefaultMimeByExtension(resource.file.name)//get content type by file extension
+            val resourceCompressible = !excludedMimeType(contentType)
+            if (!resourceCompressible)
+                acceptCompressType = CompressType.NONE
+            val resultByteArray = getStaticResourceByteArray(resource, target, acceptCompressType)
+            res.setContentLength(resultByteArray.size)
+            res.setHeader(Header.CONTENT_TYPE, contentType)
+
+            if (acceptCompressType != CompressType.NONE)
+                res.setHeader(Header.CONTENT_ENCODING, acceptCompressType.typeName)
+
+            val weakETag = resource.weakETag //jetty resource use weakETag too
+            req.getHeader(Header.IF_NONE_MATCH)?.let { etag ->
+                if (etag == weakETag) {
+                    res.status = 304
+                    return true
+                }
+            }
+            res.setHeader(Header.ETAG, weakETag)
+            resultByteArray.inputStream().copyTo(res.outputStream, 2048)
+            res.outputStream.close()
+            return true
+        }
+        return false
+    }
+
+    private fun getStaticResourceByteArray(resource: Resource, target: String, type: CompressType): ByteArray {
+        if (resource.length() > Int.MAX_VALUE)
+            throw RuntimeException("Precompression doesn't support static file size large than 2GB")
+        val key = target + type.extension
+        if (!compressedFiles.containsKey(key)) {
+            compressedFiles[key] = getCompressedByteArray(resource, type)
+        }
+        return compressedFiles[key]!!
+    }
+
+    private fun getCompressedByteArray(resource: Resource, type: CompressType): ByteArray {
+        val fileInput = resource.file.inputStream()
+        val byteArrayOutputStream = ByteArrayOutputStream()
+        val outputStream: OutputStream = when {
+            compStrat.gzip != null && type == CompressType.GZIP -> {
```

Comment:
I think it's okay to do max-level compression here. Instead of checking for `compStrat.brotli` you can check for the presence of Brotli, an apply it automatically if the library is present.

Reply:
I will do.